server/src/purchase_routes.rs

```diff
@@ -0,0 +1,124 @@
+use std::collections::HashMap;
+
+use axum::{extract::Path, Extension, Json};
+use axum_extra::extract::CookieJar;
+use entity::{
+    product,
+    purchase::{self, Entity as Purchase},
+    sea_orm, user,
+};
+use sea_orm::{
+    prelude::*, DatabaseConnection, JoinType, QueryOrder, QuerySelect, Set, TransactionTrait,
+};
+
+use crate::dtos::{BuyerGroupedPurchasesDto, PurchaseDto, SellerSummaryDto};
+use crate::errors::AppError;
+
+pub(crate) async fn seller_summary(
+    Extension(ref conn): Extension<DatabaseConnection>,
+    jar: CookieJar,
+) -> Result<Json<SellerSummaryDto>, AppError> {
+    let seller_id = crate::jwt_helpers::get_login(&jar)?;
+
+    // Sold products
+    let entities = Purchase::find()
+        .join(JoinType::InnerJoin, purchase::Relation::Product.def())
+        .filter(product::Column::Seller.eq(seller_id))
+        .filter(purchase::Column::PaidDate.is_null())
+        .order_by_desc(purchase::Column::Date)
+        .all(conn)
+        .await?;
+    let mut dtos = Vec::with_capacity(entities.len());
+    // FIXME: converting this model to DTO will trigger 3 SQL queries for each entity
+    for entity in entities {
+        dtos.push(PurchaseDto::from_entity(entity, conn).await?);
+    }
+
+    // Grouped amount due per user
+    let buyer_grouped_purchases: Vec<BuyerGroupedPurchasesDto> = dtos
+        .iter()
+        .fold(HashMap::new(), |mut acc, purchase| {
+            let buyer_id = purchase.buyer.as_ref().expect("buyer must exist").id;
+            match acc.get_mut(&buyer_id) {
+                None => {
+                    acc.insert(buyer_id, vec![purchase]);
+                }
+                Some(vec) => {
+                    vec.push(purchase);
+                }
+            };
+            acc
+        })
+        .iter()
+        .map(|(_, buyer_purchases)| {
+            let amount_due: u32 = buyer_purchases
+                .iter()
+                .map(|purchase| {
+                    purchase.quantity * purchase.unit_price.expect("purchase must have unit price")
+                })
+                .sum();
```

Comment:
Is the map of purchases only created to be then iterated once?
Instead of creating the map to vectors you could just map to the sum, no?

Another thing is to use the entry api of `HashMap` with `or_default` instead of the match where you use `get_mut`.

Reply:
Right now yes, but I'm thinking in the UI we could show the transactions grouped by user instead